src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py

```diff
@@ -222,6 +222,72 @@ def new_receiver(
         return self._channel.new_receiver(name, max_size)
 
 
+class FormulaEngine3Phase:
+    """
+    The FormulaEngine evaluates formulas and streams the results.
+
+    Use the `FormulaBuilder` to create `FormulaEngine` instances.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        phase_streams: Tuple[FormulaReceiver, FormulaReceiver, FormulaReceiver],
+    ) -> None:
+        """Create a `FormulaEngine` instance.
+
+        Args:
+            name: A name for the formula.
+            phase_streams: output streams of formula engines running per-phase formulas.
+        """
+        self._name = name
+        self._channel = FormulaChannel3Phase(self._name, self)
+        self._task = None
+        self._streams = phase_streams
+
+    async def _run(self) -> None:
+        sender = self._channel.new_sender()
+        while True:
+            try:
+                phase_1 = await self._streams[0].receive()
+                phase_2 = await self._streams[1].receive()
+                phase_3 = await self._streams[2].receive()
+                msg = Sample3Phase(
+                    phase_1.timestamp,
+                    phase_1.value,
+                    phase_2.value,
+                    phase_3.value,
+                )
+            except asyncio.CancelledError:
+                logger.exception("FormulaEngine task cancelled: %s", self._name)
+                break
+            else:
+                await sender.send(msg)
+
+    async def _stop(self) -> None:
+        """Stop a running formula engine."""
+        if self._task is None:
+            return
+        await cancel_and_await(self._task)
+
+    def new_receiver(
+        self, name: Optional[str] = None, max_size: int = 50
```

Comment:
why max_size: int = 50?
receiver has default max_size = 10.

Reply:
No,  receivers have default 50:  https://frequenz-floss.github.io/frequenz-channels-python/v0.11/reference/frequenz/channels/#frequenz.channels._broadcast.Broadcast.new_receiver

Reply:
Thanks!
Then we should not override it, here. Yes?

Reply:
I don't know, if people want to, they can override, right?

or are you talking about changing the default? my understanding is that 50 has been a good default for our use cases.  do you think there is a reason to change it?

Reply:
Maybe you should store the default somewhere, like e.g. `DEFAULT_CHANNEL_SIZE` and use it everywhere.

Reply:
or as an alternative, check here if it has been set by the user and if not don't pass it to the Channels `new_receiver()` function.

Reply:
I mean that here you override the default receiver max_size.
Why you do this?

If you need receivers default max_size, then you could do:
`def new_receiver(self, name: Optional[str] = None, max_size: Optional[int] = None) `

Reply:
Why should we expect that their buffer sizes should be coupled?  `FormulaReceiver`s are very specific to a single use case and with them,  we can fine tune the default more easily.

Broadcast `Receiver`s are more general purpose and we chose 50 randomly and it seems to work for most cases.  But by decoupling them,  we get the freedom to change the general purpose `Receiver`'s buffer size without affecting the `FormulaReceiver`.